src/shogun/kernel/SpectralMixtureKernel.cpp

```diff
@@ -0,0 +1,221 @@
+
+	return result;
+}
+
+SGVector<float64_t> CSpectralMixtureKernel::get_squared_features(CDotFeatures* features)
+{
+	SGVector<float64_t> sqr_features = SGVector<float64_t>(features->get_num_vectors());
+	
+	for (int i=0; i<sqr_features.size(); i++)
+		sqr_features[i] = features->dot(i, features, i);
+
+	return sqr_features;
+}
+
+SGMatrix<float64_t> CSpectralMixtureKernel::get_square_distance_matrix()
+{
+	CDotFeatures* dot_lhs = dynamic_cast<CDotFeatures*>(lhs);
+	CDotFeatures* dot_rhs = dynamic_cast<CDotFeatures*>(rhs);
+
+	REQUIRE(dot_lhs != NULL, "Left-hand-side features must be of type CDotFeatures");
+	REQUIRE(dot_rhs != NULL, "Right-hand-side features must be of type CDotFeatures");
+	
+	SGMatrix<float64_t> lhs_mtrx = dot_lhs->get_computed_dot_feature_matrix();
+	SGMatrix<float64_t> rhs_mtrx = dot_rhs->get_computed_dot_feature_matrix();
+
+	/* rhs^T . lhs: index (i, j) is dot product of lhs_i and rhs_j.
+	   result scaled by minus two for distance computation */
+	SGMatrix<float64_t> dot_mtrx = SGMatrix<float64_t>::matrix_multiply(rhs_mtrx, lhs_mtrx, true, false, -2.0);
+
+	SGVector<float64_t> sqr_lhs = get_squared_features(dot_lhs);
+	SGVector<float64_t> sqr_rhs = get_squared_features(dot_rhs);
+
+	for (int i=0; i<dot_mtrx.num_cols; i++)
+	{
+		/* To each column vector is added a scalar: lhs_i^2
+		   and the vector [rhs_1^2 ... rhs_n^2]^T*/
+		float64_t* col_vector_ptr = dot_mtrx.get_column_vector(i);
+
+		SGVector<float64_t>::add_scalar(sqr_lhs[i], col_vector_ptr, dot_mtrx.num_rows);
+		SGVector<float64_t>::add(col_vector_ptr, 1.0, col_vector_ptr, 1.0, sqr_rhs, dot_mtrx.num_rows);
+	}
+	
+	return dot_mtrx;
+}
+
+void CSpectralMixtureKernel::compute_weight_derivative(SGMatrix<float64_t>& squared_distances, index_t index)
+{
+	for (int i=0; i<squared_distances.num_rows; i++)
+	{
+		for (int j=0; j<squared_distances.num_cols; j++)
+		{
+			// 2*w_i*exp(-2*pi^2*dist^2*sigma_i^2) * cos(2*pi*dist*mu_i)
+			float64_t sqr_dist = squared_distances(i,j);
+			float64_t exp_factor = CMath::exp(-2*CMath::pow(CMath::PI,2)*sqr_dist*CMath::pow(m_stddeviations[index],2));
+			float64_t cos_factor = CMath::cos(2*CMath::PI*CMath::sqrt(sqr_dist)*m_means[index]);
+			squared_distances(i, j) = 2*m_weights[index]*exp_factor*cos_factor;
+		}
+	}
+}
+
+void CSpectralMixtureKernel::compute_mean_derivative(SGMatrix<float64_t>& squared_distances, index_t index)
+{
+	for (int i=0; i<squared_distances.num_rows; i++)
+	{
+		for (int j=0; j<squared_distances.num_cols; j++)
+		{
+			// -2*pi*dist*mu_i^2*exp(-2*pi^2*dist^2*sigma_i^2)*sin(2*pi*dist*mu_i)
+			float64_t sqr_dist = squared_distances(i,j);
+			float64_t outer_factor = -2*CMath::PI*CMath::sqrt(sqr_dist)*CMath::pow(m_weights[index],2);
+			float64_t exp_factor = CMath::exp(-2*CMath::pow(CMath::PI, 2)*sqr_dist*CMath::pow(m_stddeviations[index], 2));
```

Comment:
@karlnapf absolutely. I have been thinking about adding a handful of them in linalg. To make it generic I thought of having an interface
```c++
linalg::elementwise_operation<_UnaryOp>(Operand operand)
linalg::elementwise_operation<_BinaryOp>(Operand operand1, Operand operand2)
// and so
```
For example, 
```c++
// the operation
namespace impl
{
template <typename T>
struct sin {
    using return_type = float64_t;
    return_type operator(T val) { return CMath::sin(val); }
};
// specialize the above for complex128_t

// the implementation
template <enum Backend, template <class> class _UnaryOp, template <class> class _Operand, typename T>
struct elementwise_operation {
    static _Operand<typename _UnaryOp<T>::return_type> compute(_Operand<T> op) {        
        // specialized for backend, Use CPU or GPU to compute things in parallel
    }
};
}
// the wrapper
// provide separate wrappers for sin/cos/product/sq/sqrt - things like that,
//  all using the same implementation struct
template <Backend backend=linalg_traits<Core>::backend, template <class> class Matrix, typename T>
Matrix<typename sin<T>::return_type> elementwise_sin(Matrix<T> operand) {
    // not totally sure for the type thing here.. will have to check it
    return impl::elementwise_operation<backend, impl::sin, Matrix, T>(operand);
}

// usage
linalg::elementwise_sin<linalg::VIENNACL>(m);
```
and then we specialize for NATIVE and ViennaCL. For GPU, these can be done using custom OpenCL kernels. For eigen3, we can use the `array` thing.

Wrote the code just from the top of my head, it may not be like what it is. But I am sensing massive performance gain. Maybe we can provide in-place operations as well (when the scalar types agree to that).

Will work on that next week. Maybe useful for @sorig :)